File: algorithm/739. Daily Temperatures/java/DailyTemperatures.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int n = temperatures.length;
        int[] result = new int[n];
        for(int i = n-1, maxi = n-1; i >= 0; i--){
            if(temperatures[i] >= temperatures[maxi]){
                maxi = i;
            }
            else{
                int step = 1;
                while(temperatures[i+step] <= temperatures[i]){
                    step += result[i+step];
                }
                result[i] = step;
            }
        }
        return result;
    }
}
/*
// We can improve the space complexity by using the calculated result for searching next warmer temp, in stead of using a stack which uses too much space
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int n = temperatures.length;
        Deque<Integer> stack = new ArrayDeque();
        int[] result = new int[n];
        for(int i = n-1; i>=0; i--){
            if(i==n-1){
                stack.addLast(i);
                result[i] = 0;
                continue;
            }
            while(!stack.isEmpty() && temperatures[i]>=temperatures[stack.peekLast()]){
               stack.pollLast();
            }
            if(stack.isEmpty()){
                result[i] = 0;
            }
            else{
                result[i] = stack.peekLast()-i;
            }
            stack.addLast(i);
        }
        return result;
    }
}
*/

File: algorithm/739. Daily Temperatures/js/DailyTemperatures.js
/**
 * @param {number[]} temperatures
 * @return {number[]}
 */
var dailyTemperatures = function(temperatures) {
    var n = temperatures.length;
    var result = new Array(n);
    for(var i = n-1, maxi = n-1; i >= 0; i--){
        if(temperatures[i] >= temperatures[maxi]){
            result[i] = 0;
            maxi = i;
        }
        else{ // temp[i] < temp[maxi]
            // use the calculated result to find next warmer temp for cur temp
            var step = 1;
            while(temperatures[i+step] <= temperatures[i]){
                step += result[i+step];
            }
            result[i] = step;
        }
    }
    return result;
};
